### ISSUE TITLE
TypeError: cannot deepcopy this pattern object

### ISSUE THREAD

**Issue by tristan957** [mesonbuild/meson#8289]
**Describe the bug**
Trying to use <dependency>.as_link_whole() on a dependency created by my project. My project uses the CMake module to generate subprojects.

```
Traceback (most recent call last):
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/mesonmain.py", line 140, in run
    return options.run_func(options)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/msetup.py", line 253, in run
    app.generate()
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/msetup.py", line 161, in generate
    self._generate(env)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/msetup.py", line 200, in _generate
    intr.run()
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 4715, in run
    super().run()
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 519, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 544, in evaluate_codeblock
    raise e
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 537, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 560, in evaluate_statement
    return self.evaluate_if(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 642, in evaluate_if
    self.evaluate_codeblock(i.block)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 544, in evaluate_codeblock
    raise e
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 537, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 550, in evaluate_statement
    return self.function_call(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 909, in function_call
    return func(node, func_args, kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 393, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 228, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 4237, in func_subdir
    self.evaluate_codeblock(codeblock)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 544, in evaluate_codeblock
    raise e
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 537, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 560, in evaluate_statement
    return self.evaluate_if(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 642, in evaluate_if
    self.evaluate_codeblock(i.block)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 544, in evaluate_codeblock
    raise e
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 537, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 550, in evaluate_statement
    return self.function_call(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 909, in function_call
    return func(node, func_args, kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 393, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 228, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 4237, in func_subdir
    self.evaluate_codeblock(codeblock)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 544, in evaluate_codeblock
    raise e
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 537, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 550, in evaluate_statement
    return self.function_call(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 909, in function_call
    return func(node, func_args, kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 393, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 228, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 4237, in func_subdir
    self.evaluate_codeblock(codeblock)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 544, in evaluate_codeblock
    raise e
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 537, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 582, in evaluate_statement
    self.evaluate_foreach(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 830, in evaluate_foreach
    self.evaluate_codeblock(node.block)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 544, in evaluate_codeblock
    raise e
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 537, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 552, in evaluate_statement
    self.assignment(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 1219, in assignment
    value = self.evaluate_statement(node.value)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 554, in evaluate_statement
    return self.method_call(cur)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 951, in method_call
    return obj.method_call(method_name, args, kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 56, in method_call
    return method(args, kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 301, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 228, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 166, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 525, in as_link_whole_method
    new_dep = self.held_object.generate_link_whole_dependency()
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/dependencies/base.py", line 299, in generate_link_whole_dependency
    new_dep = copy.deepcopy(self)
  File "/usr/lib64/python3.6/copy.py", line 161, in deepcopy
    y = copier(memo)
  File "/home/tpartin/.local/lib/python3.6/site-packages/mesonbuild/dependencies/base.py", line 261, in __deepcopy__
    setattr(result, k, copy.deepcopy(v, memo))
  File "/usr/lib64/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib64/python3.6/copy.py", line 215, in _deepcopy_list
    append(deepcopy(a, memo))
  File "/usr/lib64/python3.6/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib64/python3.6/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib64/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib64/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib64/python3.6/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib64/python3.6/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib64/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib64/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib64/python3.6/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib64/python3.6/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib64/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib64/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib64/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib64/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib64/python3.6/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib64/python3.6/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib64/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib64/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib64/python3.6/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib64/python3.6/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib64/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib64/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib64/python3.6/copy.py", line 161, in deepcopy
    y = copier(memo)
TypeError: cannot deepcopy this pattern object
```

**To Reproduce**
I am not in a position to provide a minimal example at the moment. Soon, this project will end up open-sourced and it should be easier to create a minimally-reproducible example.

I can trigger the bug by doing the following:

```meson
# change
t_dependencies += 'static' in params and params['static'] ? hse_static_dep : hse_dep
# to
t_dependencies += 'static' in params and params['static'] ? hse_static_dep.as_link_whole() : hse_dep
```

**Expected behavior**
Not error out internally

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? native
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.) RHEL 8.2
* what Python version are you using e.g. 3.6.0
* what `meson --version` 0.56.2
* what `ninja --version` if it's a Ninja build 1.8.2

**Comment by xclaesse** [comment]
The question is what does hse_static_dep contains that cannot be copied.

**Comment by mensinda** [comment]
this may be a duplicate of #7920. Could you check with meson master again?

**Comment by tristan957** [comment]
```meson
subdir('include')
subdir('binding')
subdir('kvs')
subdir('util')
subdir('c0')
subdir('cn')
subdir('kvdb')
subdir('mpool')
if unit_test_mode
	subdir('test-support')
endif

hse_sources = [
	binding_sources,
	c0_sources,
	cn_sources,
	kvdb_sources,
	kvs_sources,
	third_party_sources,
	util_sources,
]

hse_c_args = [
	'-DCOMPNAME="kvdb"',
	'-DXXH_STATIC_LINKING_ONLY',
	'-DXXH_PRIVATE_API',
]

hse_include_directories = [
	ikvdb_includes,
	public_includes,
	third_party_includes,
	util_includes,
]

hse_dependencies = [
	curl_dep,
	libmicrohttpd_dep,
	libyaml_dep,
	m_dep,
	mpool_dep,
	pthread_dep,
	liburcu_bp_dep,
	uuid_dep,
]

hse_public_dependencies = [
	uuid_dep,
]

if profiler_dep.found()
	hse_dependencies += profiler_dep
endif

if unit_test_mode
	hse_c_args += '-DHSE_UNIT_TEST_MODE=1'

	hse_mock_include_directories = [
		c0_mock_includes,
		cn_mock_includes,
		ikvdb_mock_includes,
		kvdb_mock_includes,
		kvs_mock_includes,
		mpool_mock_includes,
		public_mock_includes,
		test_support_includes,
		test_support_mock_includes,
		util_mock_includes,
		util_internal_mock_includes,
	]

	hse_include_directories += hse_mock_include_directories

	hse_sources += [
		c0_mock_sources,
		cn_mock_sources,
		ikvdb_mock_sources,
		kvdb_mock_sources,
		kvs_mock_sources,
		mpool_mock_sources,
		public_mock_sources,
		test_support_sources,
		test_support_mock_sources,
		util_mock_sources,
		util_internal_mock_sources,
	]
endif

if unit_test_mode
	hse_c_args += '-fvisibility=default'
else
	hse_c_args += '-fvisibility=hidden'
endif

component_root_includes = include_directories('.')

hse = both_libraries(
	'hse-@0@'.format(hse_major_version),
	hse_sources,
	c_args: hse_c_args,
	include_directories: hse_include_directories,
	dependencies: hse_dependencies,
	version: meson.project_version(),
	soversion: 0,
	install: true,
)

hse_dep = declare_dependency(
	link_with: hse,
	version: meson.project_version(),
	include_directories: [
		public_includes,
	],
	dependencies: hse_public_dependencies,
)

hse_static_dep = declare_dependency(
	link_with: hse.get_static_lib(),
	version: meson.project_version(),
	include_directories: [
		public_includes,
	],
	dependencies: hse_public_dependencies,
)

pkg.generate(
	hse,
	description: 'Fast key-value storage for SSDs and persistent memory',
	filebase: 'hse-@0@'.format(hse_major_version),
	name: 'hse',
	version: meson.project_version(),
	url: 'https://github.com/hse-project/hse',
	requires: hse_public_dependencies,
	subdirs: [
		'hse-@0@'.format(hse_major_version),
	],
)
```

**Comment by tristan957** [comment]
should be a little more enlightening.

I will check with meson master.
